### PR TITLE
CLOUDSTACK-9142 Migrate VM changes xmlDesc in a safe way

### PR DESCRIPTION
**Pull request by DaanHoogland**
The problem arises when the origin hypervisor has an ip addres that ends with 1, like '10.10.10.1' and the qemu VM description is containing an address that has that as part of its address, '10.10.10.100' for instance.
now migrating to '10.10.10.10' will change both addresses in the xml description file for qemu. It is fixed and unit tests are added. I am not sure yet how to integration test this. Regression will probably work so creating a PR now.

### CODE REVIEW

**Comment by remibergsma** [comment]
@DaanHoogland Please check, unit test error:
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.cloud.hypervisor.kvm.resource.LibvirtComputingResourceTest
log4j:WARN No appenders could be found for logger (org.reflections.Reflections).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Tests run: 144, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.726 sec <<< FAILURE! - in com.cloud.hypervisor.kvm.resource.LibvirtComputingResourceTest
testMigrateCommand(com.cloud.hypervisor.kvm.resource.LibvirtComputingResourceTest)  Time elapsed: 0.013 sec  <<< ERROR!
java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.substring(String.java:1960)
```

**Comment by DaanHoogland** [comment]
looking into this, probably been working on to much at the same time today, sorry

**Comment by wido** [comment]
Looking at this test I see the problem indeed. I actually created this regression while fixing it.

One thing though, you mean that the IP of the NFS server is part of the XML desc, can you show me an example? I can't think of a way though.

The code seems good to me, but running a integration test on this is indeed hard to do.

This would btw also apply for a Ceph cluster. If a Ceph monitor has a IP which 'matches' that of the host the same situation could arrise.

**Comment by rafaelweingartner** [comment]
Hi @DaanHoogland,

I would only suggest you extracting those magic numbers at line 97 to constant variables (using some descriptive names).

I also have a doubt, 
Are we using that “@author” directive? Such as the one you have at line 26 of “LibvirtMigrateCommandWrapperTest”

BTW: I really liked the “replaceIpForVNCInDescFile” method. Very nice and descriptive method name, comprehensive java doc, test cases and the method itself is not complicated. I believe that should be our code quality goal. Congratulations !

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 1]
```diff
@@ -190,4 +196,28 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
 
         return new MigrateAnswer(command, result == null, result, null);
     }
-}
\ No newline at end of file
+
+    /**
+     * This function assumes an qemu machine desription containing a single graphics element like
```
Comment (bhaisaab) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: What is there are multiple graphics element? Can we ever hit a case where there may be more than one <graphics> element?

**Comment by bhaisaab** [comment]
Except for the concern that there may be multiple graphics element where the method would clearly fail, LGTM. Please also squash your commits.

**Comment by remibergsma** [comment]
@DaanHoogland This test is failing:

```
Test migrate VM ... === TestName: test_08_migrate_vm | Status : EXCEPTION ===
ERROR
```

Details:
```
======================================================================
ERROR: Test migrate VM
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vm_life_cycle.py", line 594, in test_08_migrate_vm
    self.vm_to_migrate.migrate(self.apiclient, migrate_host.id)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 653, in migrate
    apiclient.migrateVirtualMachine(cmd)
  File "/usr/lib/python2.7/site-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 772, in migrateVirtualMachine
    response = self.connection.marvinRequest(command, response_type=response, method=method)
  File "/usr/lib/python2.7/site-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest
    raise e
Exception: Job failed: {jobprocstatus : 0, created : u'2016-01-20T21:19:08+0000', cmd : u'org.apache.cloudstack.api.command.admin.vm.MigrateVMCmd', userid : u'95279fda-bed0-11e5-90a2-5254001daa61', jobstatus : 2, jobid : u'e7a1bb0f-39a5-4868-97c9-8fda5f8a8aa1', jobresultcode : 530, jobresulttype : u'object', jobresult : {errorcode : 530, errortext : u"org.libvirt.LibvirtException: internal error: process exited while connecting to monitor: 2016-01-20T21:19:09.852246Z qemu-kvm: Failed to start VNC server on `192.168.22.22:0,password': Failed to bind socket: Cannot assign requested address\n"}, accountid : u'95278d2a-bed0-11e5-90a2-5254001daa61'}
```

Tried the test again and it failed again. Can you investigate please?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 2]
```diff
@@ -1,4 +1,5 @@
 //
+
```
Comment (borisroman) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: Why?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 3]
```diff
@@ -48,6 +49,9 @@
 @ResourceWrapper(handles =  MigrateCommand.class)
 public final class LibvirtMigrateCommandWrapper extends CommandWrapper<MigrateCommand, Answer, LibvirtComputingResource> {
 
+    private static final String CONTENTS_WILDCARD = ".*";
+    private static final String GRAPHICS_ELEM_END = "/graphics>";
```
Comment (borisroman) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: This doesn't look correct... ?

"<" missing?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java (hunk 2 above)
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: doesn't seem intentional, will have a look

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java (hunk 3 above)
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: why would that be needed, the match is conclusive enough without it. Adding it would be purely cosmetic, wouldn't it?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java (hunk 1 above)
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: @bhaisaab not in a cloudstack generated desc file. But you are right inprinciple it could exist.

**Comment by DaanHoogland** [comment]
@wido the problem we encountered in our env is not with NFS servers but with RBD. sorry for the misdirection.

**Comment by DaanHoogland** [comment]
@remibergsma that is serious. migration is failing for us with the tested code and the test is failing with the fix. I have work to do it seems. :) Can you send me the full log for comparison, please.

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 4]
```diff
@@ -190,4 +196,28 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
-}
\ No newline at end of file
+
+    /**
+     * This function assumes an qemu machine desription containing a single graphics element like
+     *     <graphics type='vnc' port='5900' autoport='yes' listen='10.10.10.1'>
+     *       <listen type='address' address='10.10.10.1'/>
+     *     </graphics>
+     * @param xmlDesc the qemu xml description
+     * @param source the ip address to migrate from
```
Comment (borisroman) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: not used?

Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: no, will remove. tnx. it was used in the initial version.

**Comment by DaanHoogland** [comment]
@rafaelweingartner I have no idea about 'those magic numbers' I didn't touch them.
I can remove the @author. it was auto generated.
thanks for the compliment, though I myself think the javadoc could have been even more terse.

**Comment by DaanHoogland** [comment]
@remibergsma I reproduced the error. I had not copied the newlines from the problematic system into my unit test. will push a fix without hurry.

**Comment by wido** [comment]
@DaanHoogland Have you been able to fix this? Code-wise it still seems good.

**Comment by DaanHoogland** [comment]
@wido yes it was fixed. The problem was the newlines in the pattern to search for. I had not taken that in account for in the unit test. We are about to put this in production but first I have to check how the status of the tests are.

**Comment by DaanHoogland** [comment]
@remibergsma @wido can we merge this?

**Comment by davidamorimfaria** [comment]
LGTM, 4.7.1 with this fix was installed on a live environment with success. Instances can now be migrated to and from the kvm node that has an IP address which partially matches one of the ceph monitor nodes.

**Comment by swill** [comment]
It looks like @davidamorimfaria is using this in production now without issue, so that is a good sign.

@remibergsma do the tests that were previously failing now pass?

Thanks...

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 5]
```diff
@@ -190,4 +195,27 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
+     * This function assumes an qemu machine description containing a single graphics element like
+     *     <graphics type='vnc' port='5900' autoport='yes' listen='10.10.10.1'>
+     *       <listen type='address' address='10.10.10.1'/>
+     *     </graphics>
+     * @param xmlDesc the qemu xml description
+     * @param target the ip address to migrate to
+     * @return the new xmlDesc
+     */
+    String replaceIpForVNCInDescFile(String xmlDesc, final String target) {
+        final int begin = xmlDesc.indexOf(GRAPHICS_ELEM_START);
```
Comment (bhaisaab) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: While this should work for most cases, the code is not defensive. For example, it will fail for multiple graphics nodes or if there are any whitespaces between closing brackets. Consider using a dom parser.

File: plugins/hypervisors/kvm/test/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapperTest.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+//
+package com.cloud.hypervisor.kvm.resource.wrapper;
+
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+public class LibvirtMigrateCommandWrapperTest {
+    String fullfile =
+"<domain type='kvm' id='4'>\n" +
```
Comment (bhaisaab) [inline plugins/hypervisors/kvm/test/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapperTest.java]: Move this string into a separate file under resource/ for this test case. Read the file from the resource path.

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java (hunk 5 above)
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: @bhaisaab that is out of scope I think. The node is under ACS control. Adding a xml parser would be nice, I agree.

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 122
 Hypervisor xenserver
 NetworkType Advanced
 Passed=105
 Failed=13
 Skipped=4

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**

**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_deploy_vgpu_enabled_vm
test_06_copy_template
test_06_copy_iso

**Passed test suits:**
integration.smoke.test_deploy_vm_with_userdata.TestDeployVmWithUserData
integration.smoke.test_affinity_groups_projects.TestDeployVmWithAffinityGroup
integration.smoke.test_portable_publicip.TestPortablePublicIPAcquire
integration.smoke.test_over_provisioning.TestUpdateOverProvision
integration.smoke.test_global_settings.TestUpdateConfigWithScope
integration.smoke.test_scale_vm.TestScaleVm
integration.smoke.test_service_offerings.TestCreateServiceOffering
integration.smoke.test_loadbalance.TestLoadBalance
integration.smoke.test_routers.TestRouterServices
integration.smoke.test_reset_vm_on_reboot.TestResetVmOnReboot
integration.smoke.test_snapshots.TestSnapshotRootDisk
integration.smoke.test_deploy_vms_with_varied_deploymentplanners.TestDeployVmWithVariedPlanners
integration.smoke.test_network.TestDeleteAccount
integration.smoke.test_non_contigiousvlan.TestUpdatePhysicalNetwork
integration.smoke.test_deploy_vm_iso.TestDeployVMFromISO
integration.smoke.test_public_ip_range.TestDedicatePublicIPRange
integration.smoke.test_multipleips_per_nic.TestDeployVM
integration.smoke.test_regions.TestRegions
integration.smoke.test_affinity_groups.TestDeployVmWithAffinityGroup
integration.smoke.test_network_acl.TestNetworkACL
integration.smoke.test_pvlan.TestPVLAN
integration.smoke.test_volumes.TestCreateVolume
integration.smoke.test_ssvm.TestSSVMs
integration.smoke.test_nic.TestNic
integration.smoke.test_deploy_vm_root_resize.TestDeployVM
integration.smoke.test_resource_detail.TestResourceDetail
integration.smoke.test_secondary_storage.TestSecStorageServices
integration.smoke.test_vm_life_cycle.TestDeployVM
integration.smoke.test_disk_offerings.TestCreateDiskOffering

**Comment by wido** [comment]
Update on this? Is this ready to be merged? We are missing one LGTM.

Code looks good to me.

**Comment by rhtyd** [comment]
LGTM, though it would be great if @DaanHoogland can move part of the test string from the Test to an xml file under resources

**Comment by DaanHoogland** [comment]
@rhtyd tomorrow #kingsday, I will have some free time ;) loadResourceFromClassPath you are talking about, huh?

**Comment by rhtyd** [comment]
@DaanHoogland yes that or something similar so test data can be separate to its own file

**Comment by DaanHoogland** [comment]
@rhtyd as a part of loading from resource I am restructuring the project layout to adhere to maven standards. Doing so it seems not appropriate to include such a change in this bug fix.@swill I think this can be merged.

**Comment by swill** [comment]
So if I am following this correctly, we are all set.  @rhtyd requested a change to move the test data to an overridable location, but did not require this change for his LGTM.  @DaanHoogland would have to restructure the plugin in order to make that happen and feels it is outside the scope of this fix and would prefer to address that to a future improvement.  Does that all make sense?

I will put this in my merge queue, but will wait for confirmation that we are good to go before merging.  Thanks guys...

**Comment by DaanHoogland** [comment]
That's correct. I am doing the restructure now based on this PR's source branch and will make a new PR shortly. If it gets tested before this one I will close this one but it is really a pedantic play-with-myself kind of thing and the wether outside is fine.

**Comment by rhtyd** [comment]
@swill yes that was a good to have suggestion; keep large string based test data separated from test, though since this request is cosmetic in nature I'm okay if we decide to merge

**Comment by swill** [comment]
Thanks guys.  I am fine with merging it as is, but does it make sense to wait a bit to see what the discussion brings up in the #1521 PR?  Or will it not affect this PR?

**Comment by DaanHoogland** [comment]
No it won't. the fix in here is valid whichever way we go.

**Comment by swill** [comment]
Ok perfect.  I will add this to my merge queue.  Thanks for clarifying guys...
